accord-core/src/test/java/accord/impl/basic/DelayedExecutorService.java

```diff
@@ -0,0 +1,107 @@
+package accord.impl.basic;
+
+import accord.burn.random.Decision;
+import accord.burn.random.IntRange;
+import accord.burn.random.RandomInt;
+import accord.burn.random.SegmentedIntRange;
+import org.apache.cassandra.concurrent.FutureTask;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Random;
+import java.util.concurrent.AbstractExecutorService;
+import java.util.concurrent.Callable;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+
+public class DelayedExecutorService extends AbstractExecutorService
+{
+    private final PendingQueue pending;
+    private final Random random;
+    private final RandomInt jitterInNano;
+
+    public DelayedExecutorService(PendingQueue pending, Random random)
+    {
+        this.pending = pending;
+        this.random = random;
+        this.jitterInNano = new SegmentedIntRange(
+                new IntRange(microToNanos(0), microToNanos(50)),
+                new IntRange(microToNanos(50), msToNanos(5)),
+                // this is different from Apache Cassandra Simulator as this is computed differently for each executor
```

Comment:
Ideally it would probably be per-node for both, but not super important. I thought the C* simulator also used the random walk behaviour for jitter, but it looks like it doesn't. It would be good if both did, to simulate nodes becoming unhealthy for a period, then restoring behaviour. But not essential for this patch.

Reply:
> Ideally it would probably be per-node for both

makes sense, nodes coordinating to act badly at the same time isn't realistic.

> I thought the C* simulator also used the random walk behaviour for jitter, but it looks like it doesn't. It would be good if both did, to simulate nodes becoming unhealthy for a period

At least in this context I see `RandomWalkDecision` and tested and don't see such a behavior... Changed test to `RandomWalkPeriod` and see do see some logical temporal behavior but don't 100% grasp why yet; below are the results in my testing

```
# RandomWalkDecision for 10k samples
True seen 2 times in a row
True seen 2 times in a row
True seen 2 times in a row
True seen 3 times in a row
True seen 2 times in a row
True seen 2 times in a row
True seen 2 times in a row
True seen 3 times in a row
True seen 2 times in a row
True seen 2 times in a row
True seen 2 times in a row
True seen 2 times in a row
True seen 2 times in a row
True seen 2 times in a row
True seen 2 times in a row
True seen 4 times in a row
True seen 2 times in a row
True seen 2 times in a row
True seen 2 times in a row
True seen 2 times in a row
True seen 2 times in a row
True seen 2 times in a row
True seen 2 times in a row
True seen 2 times in a row
True seen 2 times in a row
True seen 2 times in a row
True seen 2 times in a row
```

```
# RandomWalkPeriod for 10k samples
Seq 82
Seq 136
Seq 18
Seq 4
Seq 2
Seq 7
Seq 3
Seq 8
Seq 3
Seq 10
Seq 34
Seq 5
Seq 15
Seq 214
Seq 13
Seq 2
Seq 6
Seq 216
Seq 243
```

The result are sequences seen where the result is > the mid point of the range, so with `new LongRange(100, 2000, MICROSECONDS, NANOSECONDS)` those are sequences seen greater than 1ms.

I do agree this type of behavior is desirable, so will take a look at this logic and see if I can get a working `Decision` that has a temporal behavior of returning `true`; ill time box it to a 1-3h though

Reply:
I created a `RandomWalkPeriodChance` that looks to be working in this manner, but the true happens 58% of the time where as we want around 1-10%... so trying to mess with this to get that involved

Reply:
In better understanding `RandomWalkDecision`, the reason for the period being so low is that we are asking for between 1-10% chance, so we can't have long periods in order to respect that... for that range here is what I see

```
Number of trues: 4110(4.109999999999999%)
Number of false: 95890(95.89%)
True seq count 0 happened 91999 times
True seq count 1 happened 3689 times
True seq count 2 happened 187 times
True seq count 3 happened 14 times
True seq count 4 happened 1 times
```

switching between 10-20% I see

```
Number of trues: 15340(15.340000000000002%)
Number of false: 84660(84.66%)
True seq count 0 happened 71748 times
True seq count 1 happened 10884 times
True seq count 2 happened 1690 times
True seq count 3 happened 285 times
True seq count 4 happened 44 times
True seq count 5 happened 9 times
```

Using `FixedChance` I see

```
Number of trues: 2909(2.9090000000000003%)
Number of false: 97091(97.09100000000001%)
True seq count 0 happened 94262 times
True seq count 1 happened 2753 times
True seq count 2 happened 72 times
True seq count 3 happened 4 times
```

Reply:
Given this, there isn't a real change between `RandomWalkDecision` and `FixedChance` at this frequency, so will exclude from this patch.

Reply:
Could I take a look at your RandomWalkPeriodChance? What I would probably do is have some chance that the RandomWalkPeriod can exceed its normal min/max, and then the min/max are flipped to the expanded range until they return to their basic range. Or something like this. But, also, no need to agonise over this.

Reply:
its in my git history, revert the last commit and you have a `main` method that I was using for my testing; edb8909f1fd8ff594f39ab1767e696d04a7d8d61 removes

Reply:
Here's my attempt: https://github.com/belliottsmith/cassandra-accord/tree/18203-suggest

This looks to produce some interesting patterns by switching between large and small ranges with some chance, and the length of a large/small period is used to determine how likely we are to switch. It appears to produce a correct distribution over a long enough sample period while still giving very lengthy periods of high jitter.

Reply:
thanks, ill take a closer look.  at attempts I see it ranges between 5% and 13%, with the avg being around 7-8%.

saw some rather crazy cases like

```
103	3
	105	2
	111	2
	118	2
	144	2
	181	2
	208	2
	211	2
	295	2
	484	2
```

so many sequences of bad back and forth; in this case 484 long times!

Ill take a closer look Monday, but with these results could drop `SegmentedIntRange` in favor of `SegmentedRandomWalkPeriod` and drop `Decision` as well.

Reply:
If you run long enough it averages to around 6.2% for me which is 1/16 as requested. But you can get some very long runs of bad luck!

We should probably randomise the ratio for each test, or randomise the range of ratios per test and the specific ratio per host

Reply:
I have been working on unit tests and finding this class is more complex than expected... I had to add a concept of "resample" so we keep adding more attempts to the loop when the target ratio isn't met... I start with 100k samples and in cases like the following, I need `21473837` resamples (`21473837 * 100 = 2,147,383,700` attempts needed).

```
int maxSmall = 78782771;
int maxLarge = 79722613;
int smallRatio = 14;
SegmentedRandomWalkPeriod period = new SegmentedRandomWalkPeriod(new IntRange(0, maxSmall), new IntRange(0, maxLarge), smallRatio, rs);
```

It seems that when it gets biased towards large it takes too many samples before getting back inline...  It also requires that `large` overlaps with `small` (though I think this is fixable), which makes this class less desirable I think; I personally feel that the existing `SegmentedIntRange` is better as it stays within the expected range.

If you want to take a look at the tests, they are in `accord.burn.random.SegmentedRandomRangeTest`.  `disjoint` tests `SegmentedIntRange` and `overlap` tests `Decision.SegmentedRandomWalkPeriod`.

Reply:
bq. It also requires that large overlaps with small (though I think this is fixable)

Why is this a problem?

bq. It seems that when it gets biased towards large it takes too many samples before getting back inline... 

I'm not convinced this is a major issue - we don't need to precisely hit the ratio every run, or even at all within a run. But, if you are concerned we could very simply bound the number of steps we permit to stay in the large side so we can call it a day.

I think we are probably overthinking this now, though, and should pick something and move to greener pastures. It is fine either way.

Reply:
Created a new `FrequentLargeRange` which has a simple way to generate "runs" and makes sure it stays within the expected range.  This gives us the benefit that we can have large runs, that we can have disjoint small/large ranges, and that the ratio is respected in real time.

Reply:
Like I said, the range isn’t something we should be strict about. We’re just generating config to generate random values, and there’s no such thing as “staying within a range” for random values - that only makes sense over some time horizon and frequency, neither of which we configure.

a random walk is independently nice because it isn’t a uniform distribution but represents shifting patterns of behaviour, not just at the small/large boundary but within each scheme.

however, we’ve agonised over this for many times longer than we should have so 🤷

Reply:
(Your class could easily use RandomWalkPeriod for small and large if you want though - but really, I’ll stop suggesting things now and we’ll take whatever you’re happy with )